common/remote/sockfile.cpp

```diff
@@ -3969,62 +4138,192 @@ class CRemoteDiskReadActivity : public CSimpleInterfaceOf<IRemoteActivity>, impl
         return makeLocalFposOffset(part, prefetchBuffer.tell());
     }
 };
 
+
 IRemoteActivity *createRemoteDiskRead(IPropertyTree &actNode)
 {
-    const char *fileName = actNode.queryProp("fileName");
-    unsigned __int64 chooseN = actNode.getPropInt64("choosen", defaultFileStreamChooseN);
-    unsigned __int64 skipN = actNode.getPropInt64("skipN", defaultFileStreamSkipN);
-    bool compressed = actNode.getPropBool("compressed");
-
-    bool inputGrouped = actNode.getPropBool("input_grouped", false);
-    bool outputGrouped = actNode.getPropBool("output_grouped", false);
-    if (!inputGrouped && outputGrouped)
-        outputGrouped = false; // perhaps should fire error
-    Owned<IOutputMetaData> inMeta = getTypeInfoOutputMetaData(actNode, "input", inputGrouped);
-    Owned<IOutputMetaData> outMeta = getTypeInfoOutputMetaData(actNode, "output", outputGrouped);
-    if (!outMeta)
+    return new CRemoteDiskReadActivity(actNode);
+}
+
+class CRemoteIndexBaseActivity : public CRemoteDiskBaseActivity
+{
+    typedef CRemoteDiskBaseActivity PARENT;
+
+protected:
+    bool isTlk = false;
+    bool allowPreload = false;
+    unsigned __int64 seeks = 0;
+    unsigned __int64 scans = 0;
+    Owned<IKeyIndex> keyIndex;
+    Owned<IKeyManager> keyManager;
+
+    void checkOpen()
+    {
+        if (opened)
+            return;
+        keyIndex.setown(createKeyIndex(fileName, 0, isTlk, allowPreload)); // JCSMORE crc
+        keyManager.setown(createLocalKeyManager(*record, keyIndex, nullptr, true));
+        filters.createSegmentMonitors(keyManager);
+        keyManager->finishSegmentMonitors();
+        keyManager->reset();
+
+        opened = true;
+    }
+    void close()
+    {
+        keyManager.clear();
+        keyIndex.clear();
+        opened = false;
+        eofSeen = true;
+    }
+public:
+    CRemoteIndexBaseActivity(IPropertyTree &config)
+    {
+        isTlk = config.getPropBool("isTlk");
+        allowPreload = config.getPropBool("allowPreload");
+    }
+};
+
+class CRemoteIndexReadActivity : public CRemoteIndexBaseActivity
+{
+    typedef CRemoteIndexBaseActivity PARENT;
+
+    unsigned __int64 chooseN = 0;
+public:
+    CRemoteIndexReadActivity(IPropertyTree &config) : PARENT(config)
+    {
+        chooseN = config.getPropInt64("chooseN", defaultFileStreamChooseNLimit);
+        inMeta.setown(getTypeInfoOutputMetaData(config, "input", false));
+        outMeta.setown(getTypeInfoOutputMetaData(config, "output", false));
+        if (!outMeta)
+            outMeta.set(inMeta);
+
+        initCommon(config);
+    }
+// IRemoteActivity impl.
+    virtual const void *nextRow(MemoryBufferBuilder &outBuilder, size32_t &retSz) override
+    {
+        if (eofSeen)
+            return nullptr;
+        checkOpen();
+        if (!eofSeen)
+        {
+            if (processed < chooseN)
+            {
+                while (keyManager->lookup(true))
+                {
+                    seeks += keyManager->querySeeks();
```

Comment:
I think this will accumulate seeks and scans multiple times.

Reply:
I would expect the call to be in serialize/change key manger, and also reset the counts at that point.

Reply:
agreed, though am not sure yet at the moment where to serialize. Should be at end of each buffer load, but will some thought to ensure fits in with existing row buffer streaming.

I will move/comment out this code for now and issue of this meta info. will be address by HPCC-19578